src/constraint/SequentialSampleable.cpp

```diff
@@ -0,0 +1,137 @@
+#include "aikido/constraint/SequentialSampleable.hpp"
+
+#include <dart/common/StlHelpers.hpp>
+
+namespace aikido {
+namespace constraint {
+
+//==============================================================================
+class SequentialSampleGenerator : public SampleGenerator
+{
+public:
+  SequentialSampleGenerator(
+      statespace::StateSpacePtr stateSpace,
+      std::vector<std::unique_ptr<SampleGenerator>> generators)
+    : mStateSpace(std::move(stateSpace))
+    , mGenerators(std::move(generators))
+    , mIndex(0)
+  {
+#ifndef NDEBUG
+    for (const auto& generator : mGenerators)
+      assert(generator->getStateSpace() == mStateSpace);
+#endif
+  }
+
+  // Documentation inherited
+  statespace::StateSpacePtr getStateSpace() const override
+  {
+    return mStateSpace;
+  }
+
+  // Documentation inherited
+  bool sample(statespace::StateSpace::State* state)
+  {
+    while (canSample())
+    {
+      if (mGenerators[mIndex]->canSample())
+      {
+        if (mGenerators[mIndex]->sample(state))
+          return true;
+      }
+      mIndex++;
+    }
```

Comment:
This logic is still inefficient because of calling `canSample()`; `canSample()` iterates over the generators remaining while `canSample()` is called repeatidly in this logic.

It seems that calling both of `canSample()` is not necessary. What would be the problem if we remove calling `canSample()` and just calling `sample(...)` iterating the generators?

Reply:
Yeah that was my concern as well. I wondered if changing `mIndex` within `canSample()` would make it any better but I guess that wouldn't be as clean or ideal.

Reply:
I don't think that's a desirable behavior.

Could you point me out what's your concern with the following proposal?:
```cpp
for (auto i = mIndex; i < mGenerators.size(); ++i)
{
  if (mGenerators[mIndex]->sample(state))
    return true;

  mIndex++;
}

return false;
```

Reply:
The current [updated] implementation has the same behavior as your proposal.

Reply:
Oh wait, actually I think it is beneficial to get rid of `canSample()`, for instance if a `SequentialSampler` is composed of more `SequentialSampler`s, the current implementation would be still inefficient.